src/tstool/tstool.c

```diff
@@ -0,0 +1,620 @@
+			if (i < num_known)
+				len = snprintf(buf, space, " %s", names[i]);
+			else
+				len = snprintf(buf, space, " %d", i);
+			if (len >= space)
+				break;
+			else
+				space -= len, buf += len;
+		}
+	}
+
+	if (flags != 0)
+		snprintf(buf + space - rewind, rewind, "%s", ellipsis);
+}
+
+static int decode_option(const char **names, size_t num_known,
+			 unsigned long *option, const char *text)
+{
+	int i;
+
+	for (i = 0; i < num_known && strcmp(text, names[i]); i++);
+	if (i == num_known) {
+		ERROR("option %s invalid\n", text);
+		return 1;
+	}
+
+	*option = i;
+	return 0;
+}
+
+
+static int do_init(void)
+{
+	int rc;
+
+	/* Start netlink service */
+	netlink = sfptpd_netlink_init();
+	if (netlink == NULL) {
+		CRITICAL("could not start netlink\n");
+		return EXIT_FAILURE;
+	}
+
+	rc = sfptpd_netlink_scan(netlink);
+	if (rc != 0) {
+		CRITICAL("scanning with netlink, %s\n",
+			 strerror(rc));
+		return rc;
+	}
+
+	/* Wait 5 seconds for initial link table */
+	initial_link_table = sfptpd_netlink_table_wait(netlink, 1, 5000);
+	if (initial_link_table == NULL) {
+		CRITICAL("could not get initial link table, %s\n",
+			 strerror(errno));
+		return EXIT_FAILURE;
+	}
+
+	/* Start clock management */
+	rc = sfptpd_clock_initialise(config, NULL);
+	if (rc != 0)
+		return EXIT_FAILURE;
+
+	/* Start interface management */
+	rc = sfptpd_interface_initialise(config, NULL,
+					 initial_link_table);
+	if (rc != 0)
+		return EXIT_FAILURE;
+
+	return 0;
+}
+
+static void do_finit(void)
+{
+	sfptpd_clock_shutdown();
+	sfptpd_interface_shutdown(config);
+	if (netlink != NULL)
+		sfptpd_netlink_finish(netlink);
+}
+
+static void usage(FILE *stream)
+{
+	fprintf(stream,
+		"syntax: %s [OPTIONS] SUBSYSTEM COMMAND..\n"
+		"\n"
+		"  OPTIONS\n"
+		"        --persistent            Use sfptpd persistent frequency adjustment\n"
+		"        --initial               Perform sfptpd initial clock correction\n"
+		"    -h, --help                  Show usage\n"
+		"    -v, --verbose               Be verbose\n\n"
+		"  CLOCK SUBSYSTEM\n"
+		"    clock list                  List clocks\n"
+		"    clock info CLOCK            Show clock information\n"
+		"    clock get CLOCK             Read clock\n"
+		"    clock step CLOCK OFFSET     Step clock\n"
+		"    clock slew CLOCK PPB        Adjust clock frequency\n"
+		"    clock set_to CLOCK1 CLOCK2  CLOCK1 := CLOCK2\n"
+		"    clock diff CLOCK1 CLOCK2    CLOCK1 - CLOCK2\n"
+		"    clock dedup                 Deduplicate shared phc devices\n\n"
+		"      CLOCK := <phcN> | <ethN> | system\n\n"
+		"  INTERFACE SUBSYSTEM\n"
+		"    interface list              List physical interfaces\n"
+		"    interface info INTF         Show interface information\n"
+		"    interface set_ts TX RX      Set timestamp modes\n\n"
+		"      INTF := <ethN>\n"
+                "      See 'info' response for available TX and RX modes\n",
+		program_invocation_short_name);
+}
+
+static int clock_command(int argc, char *argv[])
+{
+#define MAX_CLOCKS 2
```

Comment:
Should be elsewhere in file?

Reply:
It's a localised constant for an array with short scope. If we could use _C23_ yet I would write this as `constexpr`. Although `#define`s scattered through a file aren't ideal I think it would be worse to make it look like some broader constraint when it is just a n implementation detail about processing an option value.